hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/acl/iam/IamSessionPolicyResolver.java

```diff
@@ -0,0 +1,329 @@
+
+      result.addAll(stmtResults);
+    }
+
+    return result;
+  }
+
+  /**
+   * Ensures required input parameters are supplied.
+   */
+  private static void validateInputParameters(String policyJson, String volumeName,
+      AuthorizerType authorizerType) throws OMException {
+    if (StringUtils.isBlank(policyJson)) {
+      throw new OMException("The IAM session policy JSON is required", INVALID_REQUEST);
+    }
+
+    if (StringUtils.isBlank(volumeName)) {
+      throw new OMException("The volume name is required", INVALID_REQUEST);
+    }
+
+    Objects.requireNonNull(authorizerType, "The authorizer type is required");
+
+    if (policyJson.length() > MAX_JSON_LENGTH) {
+      throw new OMException("Invalid policy JSON - exceeds maximum length of " +
+          MAX_JSON_LENGTH + " characters", INVALID_REQUEST);
+    }
+  }
+
+  /**
+   * Parses IAM session policy and retrieve the statement(s).
+   */
+  private static Set<JsonNode> parseJsonAndRetrieveStatements(String policyJson) throws OMException {
+    final JsonNode root;
+    try {
+      root = MAPPER.readTree(policyJson);
+    } catch (Exception e) {
+      throw new OMException("Invalid policy JSON (most likely JSON structure is incorrect)", e, INVALID_REQUEST);
+    }
+
+    final JsonNode statementsNode = root.path("Statement");
+    if (statementsNode.isMissingNode()) {
+      throw new OMException("Invalid policy JSON - missing Statement", INVALID_REQUEST);
+    }
+
+    final Set<JsonNode> statements = new HashSet<>();
+
+    if (statementsNode.isArray()) {
+      statementsNode.forEach(statements::add);
+    } else {
+      statements.add(statementsNode);
+    }
+    return statements;
+  }
+
+  /**
+   * Parses Effect from IAM session policy and ensures it is valid and supported.
+   */
+  private static void validateEffectInJsonStatement(JsonNode statement) throws OMException {
+    final JsonNode effectNode = statement.get("Effect");
+    if (effectNode != null) {
+      if (effectNode.isTextual()) {
+        final String effect = effectNode.asText();
+        if (!"Allow".equalsIgnoreCase(effect)) {
+          throw new OMException("Unsupported Effect - " + effect, NOT_SUPPORTED_OPERATION);
+        }
+        return;
+      }
+
+      throw new OMException(
+          "Invalid Effect in JSON policy (must be a String) - " + effectNode, INVALID_REQUEST);
+    }
+
+    throw new OMException("Effect is missing from JSON policy", INVALID_REQUEST);
+  }
+
+  /**
+   * Reads a JsonNode and converts to a Set of String, if the node represents
+   * a textual value or an array of textual values.  Otherwise, returns
+   * an empty List.
+   */
+  private static Set<String> readStringOrArray(JsonNode node) {
+    if (node == null || node.isMissingNode() || node.isNull()) {
+      return Collections.emptySet();
+    }
+    if (node.isTextual()) {
+      return Collections.singleton(node.asText());
+    }
+    if (node.isArray()) {
```

Comment:
I'm not very familiar with JsonNode usage. Shall we check node.isArray() first before node.isTextual(), or node.isTextual() true means it's not an array?